### ISSUE TITLE
Replace NumPy with PyTorch for sequence generation

### ISSUE THREAD

**Issue by rlouf** [normal-computing/outlines#164]
The sequence generation code introduced in #139 uses NumPy. The idea was to provide a similar interface for every model implementation, whether they produce NumPy, JAX, TF, PyTorch, etc arrays. The downside of this approach is that it requires frequent memory transfer between GPU and CPU or fairly large arrays (batch size x vocab size) and substantially slows down the computations (see comments in #139).

I suggest we instead use PyTorch as the underlying array library. This way:

- Integration with `transformers` PyTorch models is straightforward. `Sequence` moves the token ids, attention mask, etc to `model.device`;
- Integration with llamacpp is possible: use `torch.from_numpy` in the `LlamaCPP` class to convert a NumPy array to a torch tensor without copying memory;
- Integration with models written in JAX/TF is also possible. No need to move the memory around thanks to [DLPack](https://github.com/dmlc/dlpack).

Since PyTorch's syntax is slightly different, we will need to make sure that we're using operators that do not copy array contents whenever possible. Also, some operators like `searchsorted` are not available on every backend, that's something we also need to take into consideration.
